### PR TITLE
feat(compiler): support optional chaining

### PR DESCRIPTION
**Pull request by hasanaburayyan**
Support option channing

Closes: https://github.com/winglang/wing/issues/436

- [x] PR title matches [Winglang's style guide](https://docs.winglang.io/contributors/pull_requests#how-are-pull-request-titles-formatted)
- [x] PR description explains motivation and solution
- [x] Tests added
- [x] Docs updated
- [x] Needs end-to-end tests (`pr/e2e-full` label)

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -3388,6 +3298,106 @@ impl<'a> TypeChecker<'a> {
 		}
 	}
 
+	fn resolve_variable_from_instance_type(
```
Comment (hasanaburayyan) [inline libs/wingc/src/type_check.rs:3301]: My cognative ability to name things is out of service today 😅 I needed to extract this logic to a function to recursively call it reference is `Type::Optional` I would have liked to do maybe a closure instead but to make closures recursive in rust was hella ugly.

File: examples/tests/invalid/optionals.w
[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply examples/tests/invalid/optionals.w:1]: @Chriscbr actually right now this compiles fine with no error
```js
struct A {
  val: num;
}

struct B {
  a: A;
}

struct C {
  b: B;
}

let c = C { b: B { a: A { val: 1 } } };

let val = c?.b?.a?.val;

assert(val == 1);
```

**Comment by Chriscbr** [comment]
@hasanaburayyan that's a good callout - I agree it sounds like something we should fix first. Smells like a grammar issue perhaps?

**Comment by MarkMcCulloh** [comment]
While a tiny change would be needed in the grammar to better identify which accessor is used, I think it's mostly just a modeling issue in wingc. 
Both `a.b` and `a?.b` are a `NestedIdentifier`, so the information about it being optional chaining is lost. So we'd either need to add a different ast struct or add a flag of some kind to `NestedIdentifier`. That way the type checker can use that info to provide proper diagnostics.

File: examples/tests/invalid/optionals.w
[the remark this answers is not in the document]
Comment (Chriscbr) [reply examples/tests/invalid/optionals.w:1]: That seems like a bug -- when a user writes "?." they are _expecting_ that they were holding an optional value, but it turns out it was not optional. So our compiler can give an error like `Error: expression 'c' is not an optional type, so its fields can be accessed directly (hint: 'c.b')`

[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply examples/tests/invalid/optionals.w:1]: Yea I agree I think this is a bug we just havent noticed yet. 😅

[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply examples/tests/invalid/optionals.w:1]: This case has been resolved with: https://github.com/winglang/wing/pull/2669

**Review by Chriscbr — APPROVED** [review]
🚢

**Pull request merged**